Feature/libsearch 81

cfitz opened the pull request:
https://issues.umd.edu/browse/LIBSEARCH-81

dsteelma-umd committed:
LIBSEARCH-81 inital commit for AS searcher

This is the first pass at the umd search Quick Search ArchivesSpace
Seacher. For searching.

This takes a request, passes it to the AS PUI, scrapes the results, and
displays them in the searchumd app.

https://issues.umd.edu/browse/LIBSEARCH-81

LIBSEARCH-81 Add .ruby files and some stylistic changes

This adds the .ruby-* files and makes some stylistics tweaks.

https://issues.umd.edu/browse/LIBSEARCH-81

LIBSEARCH-81 Make en.yml use more generic ArchivesSpace term

This replaces "The Archives" with the more generic and accurate
ArchivesSpace term.

https://issues.umd.edu/browse/LIBSEARCH-81

dsteelma-umd commented on .gitignore:
```diff
@@ -0,0 +1,7 @@
+.bundle/
```
Please add Gemfile.lock

dsteelma-umd commented on .rubocop.yml:
```diff
@@ -0,0 +1,63 @@
+    - 'test/dummy/**/*'
+
+  TargetRubyVersion: 2.3
+  TargetRailsVersion: 5.0
```
Please update to Rails 5.2

dsteelma-umd commented on .rubocop.yml:
```diff
@@ -0,0 +1,63 @@
+
+Naming/FileName:
+  Exclude:
+    - quick_search-library_website_searcher.gemspec
```
Please change this to quick_search-archives_space_searcher.gemspec

dsteelma-umd commented on Rakefile:
```diff
@@ -0,0 +1,34 @@
+
+RDoc::Task.new(:rdoc) do |rdoc|
+  rdoc.rdoc_dir = 'rdoc'
+  rdoc.title    = 'QuickSearch::ArchivesSpaceSearch'
```
Should probably be QuickSearchArchivesSpaceSearch

dsteelma-umd commented on app/searchers/quick_search/archives_space_searcher.rb:
```diff
@@ -0,0 +1,98 @@
+require 'nokogiri'
+
+module QuickSearch
+  # QuickSearch seacher for UMD Library Website
```
Please update this comment

dsteelma-umd commented on bin/rails:
```diff
@@ -0,0 +1,14 @@
+# installed from the root of your application.
+
+ENGINE_ROOT = File.expand_path('..', __dir__)
+ENGINE_PATH = File.expand_path('../lib/quick_search/archives_space_search/engine', __dir__)
```
I think this needs to be '../lib/quick_search/archives_space_searcher/engine'

dsteelma-umd commented on config/locales/en.yml:
```diff
@@ -0,0 +1,16 @@
+
+en:
+  archives_space_search:
+    display_name: "The Archives"
```
In config/locales/en.yml, I think the expectation was that the display name would be "Archives Space", not "The Archives". Did you have a particular reason for choosing that name?

Also, NCSU has accepted a pull request so that the "no_results_link" is basically not used anymore. So you can safely leave it as "<NO_RESULTS_LINK>" (see https://github.com/umd-lib/quick_search-swiftype_searcher/blob/develop/config/locales/en.yml)

cfitz replied:
Yeah, that's what the ticket requested. I think they want to avoid "Archivesspace" because it's jargon ( nobody that doesn't work  here knows/cares what that is ). But "The Archives" sounds odd too. 

If they're not using the no_results_link should we just remove it?

dsteelma-umd replied:
Okay -- I see where you got "The Archives" from.

There are two audiences in play here. We want the searchers to be as generic as possible, with UMD-specific configuration being done in the searchumd-special application.

The hope is that a searcher might be useful to non-UMD folks (although that is unlikely in this particular case), and so the terminology should reflect that. So I think we should definitely use "ArchivesSpace" in the searcher. If we want to change it to "The Archives" in the searchumd-special application, we can do that by simply editing the en.yml file in that application.

As for the "no_results_link" in the config/locales/en.yml file, the backstory is as follows.

In NCSU QuickSearch, there is a "loaded_link" and a "no_result_link". The "loaded_link" is intended to provide a link so that the user can go to the native search interface with the query term already filled in (so that if they want to see additional results from that search term, they just click the link). The "no_result_link" is used when no search results are found, and generally just takes the user to the native search interface page.

In NSCU QuickSearch v0.2.0 (the latest released version), the "loaded_link" is configurable either in the en.yml file, or by a "loaded_link" method in the searcher. The "no_results_link" is only configurable in the en.yml file. This is a problem when we are setting the "loaded_link" via environment variables in the .env file -- there is no way to make similar changes to the "no_results_link" value.

UMD submitted a pull request to NCSU, which they accepted, which enables a "no_results_link" property to be added to the searcher configuration file. This enables the property to be configured via the environment variables. In order to preserve backward compatibility, the QuickSearch base class still returns the "no_results_link" value (if it exists) from the en.yml file. If the "no_results_link" value exists in the en.yml file, it is always returned. The changes have not yet been released in an actual version.

The searchumd-special application is based on the NSCU QuickSearch "master" branch, so we can take advantage of this functionality. We want to use the new functionality in the "searchumd-special" application, but keep the searcher itself compatible with v0.2.0.

So we should leave the "no_results_link" in the "quick_search-archives_space_searcher/config/locales/en.yml" file.

The following changes should be made to the "searchumd-special" application:

1) add a "no_results_link" property to "searchumd-special/config/searchers/archives_space_config.yml", presumably something like:

```
defaults:
    ...
    # Link to provide when no results are found
    no_results_link: "<%= ENV['ARCHIVES_SPACE_URL'] %>"
```

Note: When I tried this, I got sent to "https://archives.lib.umd.edu/search", which displays an error. I think we want to rework things so that we have a "base url" of "https://archives.lib.umd.edu/" and a "search_url" (used for the actual searching) that includes the "/search" path. Basically when a "no results" link appears, we want to send them to the home page, not a search page that displays an error (even if they can still search from that page).

2) In the "searchumd-special" application, we should override the en.yml from the searcher by editing the "searchumd-special/config/locales/en.yml" file and adding:

```
  archives_space_search:
    no_results_link:
```
This will ensure that the value in the configuration file is used.

You can test this by running the "searchumd-special" application and putting in a nonsensical search term such as "foobarbazquuz". The search will return no results and the "Try a different search..." link should go to the ArchivesSpace home page.

dsteelma-umd commented on quick_search-archives_space_searcher.gemspec:
```diff
@@ -0,0 +1,22 @@
+Gem::Specification.new do |s|
+  s.name        = 'quick_search-archives_space_searcher'
+  s.version     = ArchivesSpaceSearcher::VERSION
+  s.authors     = ['Chris Fitzpatrick']
```
Please update to reflect "UMD Libraries" authorship, email address and home page (see https://github.com/umd-lib/quick_search-swiftype_searcher/blob/develop/quick_search-swiftype_searcher.gemspec)

dsteelma-umd commented on README.md:
```diff
@@ -1,2 +1,27 @@
+and copy the "config/archives_space_config.yml" file
+in this gem into it. Follow the instructions in the file to configure the
+searcher.
+
```
For consistency with the other searchers, please indicate the "render_module" stanza to use in the QuickSearch application. See https://github.com/umd-lib/quick_search-swiftype_searcher

dsteelma-umd commented on app/searchers/quick_search/archives_space_searcher.rb:
```diff
@@ -0,0 +1,98 @@
+    end
+
+    def base_query_params
+      QuickSearch::Engine::ARCHIVESSPACE_CONFIG['query_params']
```
For naming consistency, please rename "ARCHIVESSPACE_CONFIG" to "ARCHIVES_SPACE_CONFIG" in this file and in config/initializers/archives_space_searcher_config.rb

dsteelma-umd commented on config/initializers/archives_space_searcher_config.rb:
```diff
@@ -0,0 +1,11 @@
+  File.expand_path('../archives_space_config.yml', __dir__)
+].first { |file| File.exists? file }
+
+QuickSearch::Engine::ARCHIVESSPACE_CONFIG =
```
For naming consistency, please rename "ARCHIVESSPACE_CONFIG" to "ARCHIVES_SPACE_CONFIG" in this file and in app/searchers/quick_search/archives_space_searcher.rb

dsteelma-umd commented on config/archives_space_config.yml:
```diff
@@ -0,0 +1,38 @@
+    from_year: [] 
+    to_year: []
+    q: ''
+
```
There should be a "loaded_link" parameter that points to the "native" search interface (and populates the native search interface with the query term) , and a "no_results_link" parameter that points to the native search interface page.

cfitz replied:
Hm, I don't see where these settings are used in this searcher? Is it required?

dsteelma-umd replied:
See above comment. Because you have a "loaded_link" method in the searcher, you don't need a "loaded_link" parameter in config/archives_space_config.yml. Also, we'll add the "no_results_link" parameter in the search application, so no changes are needed for this file.

dsteelma-umd reviewed (commented):
1) Please add .ruby-gemset and .ruby-version files. To be consistent with the other searchers (and the searchumd-special application), the Ruby version should be v2.3.7. The gemset name should be the same as the GitHub repository (quick_search-archives_space_searcher).

2) This should have been more explicit in the issue description -- please create a pull request against the "umdsearch-special" repository (https://github.com/umd-lib/searchumd-special), modifying the application to include this searcher, replacing the "arXiv" searcher. This should include using the "dotenv" gem and adding to the "env_example" file to parameterize the parameters (such as "search_url") needed for the searcher.

A sample pull request (against the "searchumd" application, not "searchumd-special) for adding a searcher is:

* https://github.com/umd-lib/searchumd/pull/41

The Git commits and pull request can use this issue for its Jira issue number.

3) Please fix Rubocop offenses.

4) For consistency with the other searchers, the module name in lib/quicksearch/archives_space_searcher/engine.rb file should probably be "QuickSearchArchivesSpaceSearcher" not "ArchivesSpaceSearcher". This will necessitate changes in other files.

I think NSCU's original intention was that things would be set up at

```
module QuickSearch
  module ArchivesSpaceSearcher
```

(at least that's what their Rails generator generates). However, this causes a name conflict, and so we compensated by changing the name to "QuickSearchArchivesSpaceSearcher". I think the issue the NCSU QuickSearch is having is fixable (see https://issues.umd.edu/browse/LIBSEARCH-17 for an approach that I think _might_ work).

So if/when this issue gets fixed, we can simply go back and update the searchers.

I'm willing to be convinced that this is a bad idea if you feel strongly about it.

The following are FYI:

1) Instead of pushing up the feature branch into the repository, it is okay to specify that the reviewer should clone the repository, and checkout the pull request, then modify the "Gemfile" in the "searchumd-special" as:

```
gem 'quick_search-archives_space_searcher',
    path: '../quick_search-archives_space_searcher'
```

this keeps the Git forking/branching consistent with other projects. See LIBSEARCH-71 (https://issues.umd.edu/browse/LIBSEARCH-71) for an example.

2) Lately we've been tending to put the verification steps in the Jira issue description, instead of in a comment. This makes it easier to edit/modify as necessary, and makes it easy to find (we've had some Jira issues with lots of comments). The guidance we are now providing to the new GAs is to treat comments as a short-term "chat" about the issue, while the issue description should be edited as needed to include information useful in the long term.

cfitz commented:
Hi, 

Ok, I think I got most things updated. I was using the "Customizing Searchers" documentation in the QuickSearch repo which used a rails generator to make the engine, so I didn't actually even notice some of this stuff.

I didn't get any rubocop errors, but I think I might have run a newer version. Is there a reason to use an older version? It's gotten less nit-picky, so newer versions are much easier to use.

About the naming stuff, yeah, that's a departure from the Rails conventions. I think QuickSearchArchivesSpaceSearcher sounds really redundant and java-ish, but changing everything else seems like more work than just pasting "QuickSearch" to a few files, so I updated it.

( Also, "ArchivesSpace" is one word, so the _ in-between is not really needed. ) 

I put comments in above about a couple things. Let me know if I need to change anything else.

dsteelma-umd commented:
The only reason for using the particular Rubocop version is to be consistent with the other searchers.

As for "ArchivesSpace" -- we've run into this before with "LibGuides" (and "WorldCat") and we went with putting an underscore in so that the Rails conventions would work with us instead of against us.

cfitz commented:
Ok, so ,all that seems to need to be changed now is change on this PR is to change "The Archives" to "ArchivesSpace" in the en.yml file? 

And yes, I thing it's fine to have the _ in the file name and in names that are going to be converted to classes, but I meant for something like a constant (  ARCHIVES_SPACE_CONFIG ), you're losing the camel case aspect...upcasing ArchivesSpace is ARCHIVESSPACE and Archives Space would be ARCHIVES_SPACE.

I would recommend just going with the latest rubocop. It's way less annoying.

dsteelma-umd commented:
It depends. The other thing we want to do is to enable the "loaded_link" to point at a URL like "https://archives.lib.umd.edu/search" and the "no_results_link" to point at "https://archives.lib.umd.edu/". So I suspect there might need to be some reworking of the parameters so that we can specify a "base_url" that points to the ArchivesSpace server, and a "search_url" that appends to the base_url and is used by the "loaded_link" method.

cfitz commented:
Ok, so I just need to add "loaded_link" and "no_results_link" to the config file?

dsteelma-umd commented:
Okay, as I look more carefully at it, I realize that you can implement the "no_results" link without having to change the searcher code.

You do not need a "loaded_link" parameter in the searcher configuration, because you have a "loaded_link" method in the searcher.

In the pull request for "searchumd-special", the "search_url" parameter is configured via the "ARCHIVES_SPACE_URL" environment variable, which you have set to "https://archives.lib.umd.edu/search" (see search-umd-special/env_example).

Based on my testing, it seems that the correct URL for the "no_results_link" is the home page of the "ArchivesSpace" application -- https://archives.lib.umd.edu/. If this doesn't seem correct to you, please let me know.

So for the "searchumd-special" application we need to be able to generate two different URLs:

* https://archives.lib.umd.edu/search - used by the "loaded_link" method
* https://archives.lib.umd.edu/ - use by the "no_results_link"

The only environment variable we have to configure is "ARCHIVES_SPACE_URL", which is currently set to "https://archives.lib.umd.edu/search", and so we do not have a straightforward way of generating the "no_results_link".

What I would suggest is the following (all changes are in the "searchumd-special" application):

The "ARCHIVES_SPACE_URL" in the "env_example" file should be the URL of the ArchivesSpace server. However, we don't want to put a particular value in there, because it changes between the dev/stage/production servers.

So notionally the "ARCHIVES_SPACE_URL" should be "https://archives.lib.umd.edu/" not "https://archives.lib.umd.edu/search". In the "env_example" file in searchumd-special, though, I would expect to just see:

```
# -- config/searchers/archives_space_config.yml
# The URL of the ArchivesSpace server
ARCHIVES_SPACE_URL=
```

In "searchumd-special/config/searchers/archives_space_config.yml", we need to edit the "search_url" property to be:

```
defaults: &defaults
  search_url: "<%= ENV['ARCHIVES_SPACE_URL'] %>/search"
```

so that the "loaded_link" method gets the correct URL and path.

The "no_results_link" property in "searchumd-special/config/searchers/archives_space_config.yml" is then just:

```
defaults: &defaults
  ...
  no_results_link: "<%= ENV['ARCHIVES_SPACE_URL'] %>"
```

These changes should only be made in the "searchumd-special" application. They are not needed in the "quick_search-archives_space_searcher" code because we don't want to mandate the use of the "dotenv" gem.

Apologies for the confusion.

dsteelma-umd commented:
To summarize all of the above:

In "quick_search-archives_space_searcher":

* Update the "quick_search-archives_space_searcher/config/locales/en.yml" file to use "ArchivesSpace" terminology.

In "searchumd-special", update the following files as indicated in the previous comment:

* searchumd-special/env_example - see previous comment
* searchumd-special/config/searchers/archives_space_config.yml - see previous comment
* searchumd-special/config/locales/en.yml  - Add the following to suppress the "no_results_link" entry from the searcher:

```
  archives_space_search:
    no_results_link:
```

cfitz committed:
LIBSEARCH-81 Point search to ArchivesSpace Solr

This search runs a query agains the ArchivesSpace solr. It takes some of
the default params passed to Solr from the PUI.

https://issues.umd.edu/browse/LIBSEARCH-81
